okio/src/main/java/okio/Buffer.java

```diff
@@ -782,10 +794,23 @@ public void clear() {
   }
 
   @Override public Buffer writeString(String string, Charset charset) {
+    return writeString(string, 0, string.length(), charset);
+  }
+
+  @Override
+  public Buffer writeString(String string, int beginIndex, int endIndex, Charset charset) {
     if (string == null) throw new IllegalArgumentException("string == null");
+    if (beginIndex < 0) throw new IllegalAccessError("beginIndex < 0: " + beginIndex);
+    if (endIndex < beginIndex) {
+      throw new IllegalArgumentException("endIndex < beginIndex: " + endIndex + " < " + beginIndex);
+    }
+    if (endIndex > string.length()) {
+      throw new IllegalArgumentException(
+          "endIndex > string.length: " + endIndex + " > " + string.length());
+    }
     if (charset == null) throw new IllegalArgumentException("charset == null");
     if (charset.equals(Util.UTF_8)) return writeUtf8(string);
-    byte[] data = string.getBytes(charset);
+    byte[] data = string.substring(beginIndex, endIndex).getBytes(charset);
```

Comment:
It does on JDK 8+. Not sure about Android. I originally had a "TODO see if we can get a subset of bytes rather than bytes of substring" but deleted it before making the PR.